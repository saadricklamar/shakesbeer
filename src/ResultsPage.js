import React, { Component } from 'react';
import './ResultsPage.scss';
import Controls from './Controls.js'
import BreweryList from './BreweryList.js'
import logo from './images/shakesbeerlogosmall.png';

let starredData = JSON.parse(localStorage.getItem('userStarredList')) || [];

class ResultsPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      stateBreweries: [],
      filteredBreweries: [],
      starredBreweries: starredData,
      breweryCities: [],
      beerStyles: [],
      beerIbus: [],
      citySelection: 'All', 
      styleSelection: 'All',
      ibuSelection: 'All',
      viewingStarred: false
    }

    this.updateStarredList = this.updateStarredList.bind(this);
  }

  componentWillMount() {
    this.getStateBreweries();
  }

  getStateBreweries = () => {
    let stateBreweries = this.props.dataset.filter(brewery => {
      return brewery.state === this.props.selectedState;
    })
    this.setState({ stateBreweries: stateBreweries }, () => {
      this.refreshBreweryList();
      this.getCities();
      this.getStyles();
      this.getIbus();
    })
  }

  getCities = () => {
    let cities = this.state.stateBreweries.reduce((acc, brewery) => {
      if(!acc.includes(brewery.city)) {
        acc.push(brewery.city)
      }
      return acc;
    }, []).sort();
    this.setState({ breweryCities: cities });
  } 

  getStyles = () => {
    let styles = this.state.stateBreweries.reduce((acc, brewery) => {
      brewery.beers.forEach(beer => {
        if(!acc.includes(beer.style) && beer.style !== '') {
          acc.push(beer.style)
        }
      })
      return acc;
    }, []).sort();

    this.setState({ beerStyles: styles });
  } 

  getIbus = () => {
    let ibus = this.state.stateBreweries.reduce((acc, brewery) => {
      brewery.beers.forEach(beer => {
        const ibuMinRange = Math.round(beer.ibu / 10) * 10;
        const ibuOption = `${ibuMinRange}-${ibuMinRange + 9}`;
        if(!acc.includes(ibuOption) && beer.ibu !== null) {
          acc.push(ibuOption);
        }
      })
      return acc;
    }, []).sort().sort((a, b) => a.length - b.length);
    this.setState({ beerIbus: ibus });
  }

  updateFilterSelection = (filterName, value) => {
    const propertyName = `${filterName}Selection`;
    this.setState({ [propertyName]: value });
    this.refreshBreweryList();
  }

  toggleStarView = () => {
    this.setState({ viewingStarred: !this.state.viewingStarred }, () => {
      this.refreshBreweryList();
    });
  }

  refreshBreweryList = () => {
    this.hideBeerInfo('dropdown');
    this.hideBeerInfo('description');
    this.setState({ filteredBreweries: [...this.state.stateBreweries] }, () => this.filterByCity());
  }

  hideBeerInfo = (element) => {
    document.querySelectorAll(`.beer-${element}`).forEach(item => {
      item.classList.add('hidden');
    });
  }

  filterByCity = () => {
    if (this.state.citySelection !== 'All') {
      let filterResults = this.state.filteredBreweries.filter(brewery => {
        return brewery.city === this.state.citySelection;
      });
      this.setState({ filteredBreweries: filterResults }, () => this.filterByStyle());
    } else {
      this.filterByStyle();
    }
  }

  filterByStyle = () => {
    if (this.state.styleSelection !== 'All') {
      let filterResults = this.state.filteredBreweries.reduce((acc, brewery) => {
        let breweryStyles = brewery.beers.map(beer => beer.style);
        if(breweryStyles.includes(this.state.styleSelection)) {
          acc.push(brewery);
        }
        return acc;
      }, [])
      this.setState({filteredBreweries: filterResults}, () => this.filterByIbu());
    } else {
      this.filterByIbu();
    }
  }

  filterByIbu = () => {
    if (this.state.ibuSelection !== 'All') {
      let filterResults = this.state.filteredBreweries.reduce((acc, brewery) => {
        let breweryIbus = brewery.beers.map(beer => {
          return `${Math.round(beer.ibu / 10) * 10}-${(Math.round(beer.ibu / 10) * 10) + 9}`;
        });
        if (breweryIbus.includes(this.state.ibuSelection)) {
          acc.push(brewery);
        }
        return acc;
      }, []);
      this.setState({ filteredBreweries: filterResults }, () => this.filterByStarred());
    } else {
      this.filterByStarred();
    }
  }

  updateStarredList = (id, change) => {
    let list = this.state.starredBreweries;
    if (change === 'add') {
      list.push(id)
    } else {
      const matchingIndex = list.indexOf(list.find(obj => obj.id === id));
      list.splice(matchingIndex, 1);
    }
    this.setState({starredBreweries: list}, () => {
      localStorage.setItem('userStarredList', JSON.stringify(this.state.starredBreweries))
    });
  }

  filterByStarred = () => {
    if (this.state.viewingStarred === true) {
      let filterResults = this.state.filteredBreweries.filter(brewery => {
        return this.state.starredBreweries.includes(brewery.FIELD1);
      })
      this.setState({filteredBreweries: filterResults})
    }
  }

  goBackHome = () => {
    window.location.reload()
  }
 
  render() {
    const beerCount = this.state.filteredBreweries.reduce((sum, brewery) => {
      sum += brewery.beers.length;
      return sum;
    }, 0)    
    return (
      <div className='results-page'>
        <header>
          <i className='fas fa-arrow-left' onClick={this.goBackHome}></i>
          <img className='logo' src={logo} alt='shakesbeer logo'/>
          <h1 className='results-header'>ShakesBeer</h1>
        </header>
        <main>
          <div className='brewery-list-header'>
            <img className='flag' src={require(`./images/${this.props.selectedState.toLowerCase()}.png`)} alt='state flag'/>
            <h2 className='state-subheading'>{this.props.selectedState} Brewery List</h2> 
            <img className='flag' src={require(`./images/${this.props.selectedState.toLowerCase()}.png`)} alt='state flag'/>
          </div>
          <div className='brew-cards'>
            <Controls breweryCities={this.state.breweryCities} 
                      beerStyles={this.state.beerStyles}
                      beerIbus={this.state.beerIbus}
                      updateFilterSelection={this.updateFilterSelection}
                      toggleStarView={this.toggleStarView}
            />
            <p className='results-count'>Viewing {this.state.filteredBreweries.length} breweries with {beerCount} beers</p>
            <BreweryList filteredBreweries={this.state.filteredBreweries} 
                         dataset={this.props.dataset}
                         updateStarredList={this.updateStarredList}
                         starredBreweries={this.state.starredBreweries}
            />
          </div>
        </main>
      </div>
    );
  }
}

export default ResultsPage;