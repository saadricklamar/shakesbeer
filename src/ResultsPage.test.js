import React from 'react';
import ResultsPage from './ResultsPage.js';
import { shallow } from 'enzyme';

const mockArray = [1,2,3]
const mockData = [
  {
  "FIELD1": 90,
  "name": "The Lion Brewery",
  "city": "Wilkes-Barre",
  "state": "Pennsylvania",
  "beers": [ {
      "FIELD1": 43,
      "abv": 0.045,
      "ibu": 18,
      "id": 2544,
      "name": "Send Help",
      "style": "American Blonde Ale",
      "brewery_id": 60,
      "ounces": 12
    },
    {
      "FIELD1": 44,
      "abv": 0.055,
      "ibu": null,
      "id": 2324,
      "name": "Cast Iron Oatmeal Brown",
      "style": "American Brown Ale",
      "brewery_id": 60,
      "ounces": 12
    },
    {
      "FIELD1": 45,
      "abv": 0.06,
      "ibu": null,
      "id": 2288,
      "name": "Reprise Centennial Red",
      "style": "American Amber / Red Ale",
      "brewery_id": 60,
      "ounces": 12
    },
    {
      "FIELD1": 46,
      "abv": 0.055,
      "ibu": null,
      "id": 2287,
      "name": "Alter Ego",
      "style": "American Black Ale",
      "brewery_id": 60,
      "ounces": 12
    }
  ]
  }
]

const mockSelectedState = 'Colorado'
const mockFunc = jest.fn()

describe('ResultsPage', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow (
          <ResultsPage selectedState={mockSelectedState}
                       dataset={mockData}
          />
      )
  });

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should have default state', () => {
    expect(wrapper.state()).toEqual(
      {
        stateBreweries: [],
        filteredBreweries: [],
        starredBreweries: [],
        breweryCities: [],
        beerStyles: [],
        beerIbus: [],
        citySelection: 'All', 
        styleSelection: 'All',
        ibuSelection: 'All',
        viewingStarred: false
      }
    )
  });

  it('Should setState for breweryCities', () => {
    wrapper.setState({stateBreweries: mockData});
    expect(wrapper.state('breweryCities')).toEqual([]);
    wrapper.instance().getCities();
    expect(wrapper.state('breweryCities')).toEqual(['Wilkes-Barre']);
  });




});