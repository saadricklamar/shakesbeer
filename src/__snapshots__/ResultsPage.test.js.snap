// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsPage Should match the snapshot 1`] = `
<div
  className="results-page"
>
  <header>
    <i
      className="fas fa-arrow-left"
      onClick={[Function]}
    />
    <img
      alt="shakesbeer logo"
      className="logo"
      src="shakesbeerlogosmall.png"
    />
    <h1
      className="results-header"
    >
      ShakesBeer
    </h1>
  </header>
  <main>
    <div
      className="brewery-list-header"
    >
      <img
        alt="state flag"
        className="flag"
        src="colorado.png"
      />
      <h2
        className="state-subheading"
      >
        Colorado
         Brewery List
      </h2>
      <img
        alt="state flag"
        className="flag"
        src="colorado.png"
      />
    </div>
    <div
      className="brew-cards"
    >
      <Controls
        beerIbus={Array []}
        beerStyles={Array []}
        breweryCities={Array []}
        toggleStarView={[Function]}
        updateFilterSelection={[Function]}
      />
      <p
        className="results-count"
      >
        Viewing 
        0
         breweries with 
        0
         beers
      </p>
      <BreweryList
        dataset={
          Array [
            Object {
              "FIELD1": 90,
              "beers": Array [
                Object {
                  "FIELD1": 43,
                  "abv": 0.045,
                  "brewery_id": 60,
                  "ibu": 18,
                  "id": 2544,
                  "name": "Send Help",
                  "ounces": 12,
                  "style": "American Blonde Ale",
                },
                Object {
                  "FIELD1": 44,
                  "abv": 0.055,
                  "brewery_id": 60,
                  "ibu": null,
                  "id": 2324,
                  "name": "Cast Iron Oatmeal Brown",
                  "ounces": 12,
                  "style": "American Brown Ale",
                },
                Object {
                  "FIELD1": 45,
                  "abv": 0.06,
                  "brewery_id": 60,
                  "ibu": null,
                  "id": 2288,
                  "name": "Reprise Centennial Red",
                  "ounces": 12,
                  "style": "American Amber / Red Ale",
                },
                Object {
                  "FIELD1": 46,
                  "abv": 0.055,
                  "brewery_id": 60,
                  "ibu": null,
                  "id": 2287,
                  "name": "Alter Ego",
                  "ounces": 12,
                  "style": "American Black Ale",
                },
              ],
              "city": "Wilkes-Barre",
              "name": "The Lion Brewery",
              "state": "Pennsylvania",
            },
          ]
        }
        filteredBreweries={Array []}
        starredBreweries={Array []}
        updateStarredList={[Function]}
      />
    </div>
  </main>
</div>
`;
